ASM wrappers produce code that violates borrow rules

With the following ARM assembly function in `test.c` (found in the ARM CMSIS standard library as used in RIOT-OS; disclaimer: I have no clue what this actually does, it looks like an inline wrapper around some CPU instruction), the resulting Rust code is not buildable:

```c
#include <stdint.h>

uint64_t __SMLALDX (uint32_t op1, uint32_t op2, uint64_t acc)
{
  union llreg_u{
    uint32_t w32[2];
    uint64_t w64;
  } llr;
  llr.w64 = acc;

  asm volatile ("smlaldx %0, %1, %2, %3" : "=r" (llr.w32[1]), "=r" (llr.w32[0]): "r" (op1), "r" (op2) , "0" (llr.w32[1]), "1" (llr.w32[0]) );

  return llr.w64;
}
```

After manually fixing #306 by s/asm/llvm_asm/ on c2rust-lib.rs and test.rs, the error can be shown with:

```shell
$ c2rust transpile compile_commands.json --emit-build-file
Transpiling test.c
$ cargo +nightly build
   Compiling c2rust_out v0.0.0 (/tmp/asm)
warning: unused `#[macro_use]` import
  --> c2rust-lib.rs:13:1
   |
13 | #[macro_use]
   | ^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

error[E0499]: cannot borrow `llr.w32[_]` as mutable more than once at a time
  --> test.rs:19:18
   |
17 |     let fresh0 = &mut llr.w32[1 as libc::c_int as usize];
   |                  --------------------------------------- first mutable borrow occurs here
18 |     let fresh1;
19 |     let fresh2 = &mut llr.w32[0 as libc::c_int as usize];
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ second mutable borrow occurs here
...
24 |          "r" (op2), "0" (c2rust_asm_casts::AsmCast::cast_in(fresh0, fresh4)),
   |                                                             ------ first borrow later used here

error[E0503]: cannot use `llr.w32[_]` because it was mutably borrowed
  --> test.rs:21:18
   |
17 |     let fresh0 = &mut llr.w32[1 as libc::c_int as usize];
   |                  --------------------------------------- borrow of `llr.w32[_]` occurs here
...
21 |     let fresh4 = llr.w32[1 as libc::c_int as usize];
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ use of borrowed `llr.w32[_]`
...
24 |          "r" (op2), "0" (c2rust_asm_casts::AsmCast::cast_in(fresh0, fresh4)),
   |                                                             ------ borrow later used here

error[E0503]: cannot use `llr.w32[_]` because it was mutably borrowed
  --> test.rs:22:18
   |
17 |     let fresh0 = &mut llr.w32[1 as libc::c_int as usize];
   |                  --------------------------------------- borrow of `llr.w32[_]` occurs here
...
22 |     let fresh5 = llr.w32[0 as libc::c_int as usize];
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ use of borrowed `llr.w32[_]`
23 |     llvm_asm!("smlaldx $0, $1, $2, $3" : "=r" (fresh1), "=r" (fresh3) : "r" (op1),
24 |          "r" (op2), "0" (c2rust_asm_casts::AsmCast::cast_in(fresh0, fresh4)),
   |                                                             ------ borrow later used here

error: aborting due to 3 previous errors; 1 warning emitted

Some errors have detailed explanations: E0499, E0503.
For more information about an error, try `rustc --explain E0499`.
error: could not compile `c2rust_out`

To learn more, run the command again with --verbose.
```

I've deliberately not passed the `--target thumbv7em-none-eabihf` in to cargo as that'd necessitate *several* other workarounds to get things running, among them #297 and #307 as well as adding panic stuff.

To explore possibilities, I've altered the fresh variables to be `*mut T` rather than `&mut T`, and only reborrow for any use in the cast_in / cast_out functions. That eventually (ie. with all the no_std workarounds and LTO) the same four instructions as plain compiling clang (cf. [godbolt output](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,j:1,lang:___c,selection:(endColumn:1,endLineNumber:15,positionColumn:1,positionLineNumber:15,selectionStartColumn:1,selectionStartLineNumber:15,startColumn:1,startLineNumber:15),source:'%23include+%3Cstdint.h%3E%0A%0Auint64_t+__SMLALDX+(uint32_t+op1,+uint32_t+op2,+uint64_t+acc)%0A%7B%0A++union+llreg_u%7B%0A++++uint32_t+w32%5B2%5D%3B%0A++++uint64_t+w64%3B%0A++%7D+llr%3B%0A++llr.w64+%3D+acc%3B%0A%0A++asm+volatile+(%22smlaldx+%250,+%251,+%252,+%253%22+:+%22%3Dr%22+(llr.w32%5B1%5D),+%22%3Dr%22+(llr.w32%5B0%5D):+%22r%22+(op1),+%22r%22+(op2)+,+%220%22+(llr.w32%5B1%5D),+%221%22+(llr.w32%5B0%5D)+)%3B%0A%0A++return+llr.w64%3B%0A%7D%0A'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:armv7-cclang-10,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'1',trim:'1'),fontScale:14,j:1,lang:___c,libs:!(),options:'-target+arm-none-eabi+-mcpu%3Dcortex-m4+-mlittle-endian+-mthumb+-O3',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'armv7-a+clang+10.0.0+(Editor+%231,+Compiler+%231)+C',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)).

Might it be suitable to generally go for raw pointers rather than references when building code for ASM in the first place? While C2Rust might manage to recognize some patterns where references can work, I figure (from my very limited knowledge of assembly) it's hard to tell in general when borrow semantics can be applied and when not.

--- a/c2rust-transpile/src/translator/assembly.rs
+++ b/c2rust-transpile/src/translator/assembly.rs
@@ -883,7 +883,8 @@ impl<'c> Translation<'c> {
                     // c2rust-ast-exporter added it (there's no gcc equivalent);
                     // in this case, we need to do what clang does and pass in
                     // the operand by-address instead of by-value
-                    out_expr = mk().mutbl().borrow_expr(out_expr);
+                    self.use_feature("raw_ref_op");
+                    out_expr = mk().mutbl().raw_borrow_expr(out_expr);
                 }
 
                 if let Some(_tied_operand) = tied_operands.get(&(output_idx, true)) {
@@ -895,12 +896,13 @@ impl<'c> Translation<'c> {
                     // type conversions into the `c2rust-asm-casts` crate,
                     // so we call into that one from here.
 
-                    // Convert `x` into `let freshN = &mut x; *x`
+                    // Convert `x` into `let freshN = &raw mut x; *x`
+                    self.use_feature("raw_ref_op");
                     let output_name = self.renamer.borrow_mut().fresh();
                     let output_local = mk().local(
                         mk().ident_pat(&output_name),
                         None,
-                        Some(mk().mutbl().borrow_expr(out_expr)),
+                        Some(mk().mutbl().raw_borrow_expr(out_expr)),
                     );
                     stmts.push(mk().local_stmt(Box::new(output_local)));
 
@@ -924,7 +926,8 @@ impl<'c> Translation<'c> {
                 let mut in_expr = in_expr.into_value();
 
                 if operand.mem_only {
-                    in_expr = mk().borrow_expr(in_expr);
+                    self.use_feature("raw_ref_op");
+                    in_expr = mk().raw_borrow_expr(in_expr);
                 }
                 if let Some(tied_operand) = tied_operands.get(&(input_idx, false)) {
                     self.use_crate(ExternCrate::C2RustAsmCasts);

--- a/c2rust-transpile/tests/snapshots.rs
+++ b/c2rust-transpile/tests/snapshots.rs
@@ -423,6 +423,17 @@ fn test_volatile() {
 // arch-specific
 
 #[test]
+fn test_asm() {
+    #[cfg(target_arch = "x86_64")]
+    transpile("asm.c")
+        .arch_specific(true)
+        .expect_unresolved_import("c2rust_asm_casts")
+        .run();
+    #[cfg(not(target_arch = "x86_64"))]
+    transpile("asm.c").arch_specific(true).run();
+}
+
+#[test]
 fn test_spin() {
     transpile("spin.c").arch_specific(true).run();
 }

--- a/c2rust-transpile/tests/snapshots/arch-specific/asm.c
+++ b/c2rust-transpile/tests/snapshots/arch-specific/asm.c
@@ -1,0 +1,23 @@
+#ifdef __aarch64__
+
+// nothing here yet...
+
+#endif
+
+#ifdef __x86_64__
+
+// Tests https://github.com/immunant/c2rust/issues/398
+int six(void) {
+    int out = 0;
+    int six = 6;
+
+    asm (
+        "add %0, %1\n\t"
+        : "=r"(out)
+        : "r"(six), "0"(out)
+    );
+
+    return out;
+}
+
+#endif

